### PR TITLE
Added PFlog support

### PR DESCRIPTION
**Pull request by alexmozzhakov**
Enhancement

## What does this PR do?

Adds support for sniffing PFLog packets in Packetbeat

## Why is it important?

PFlog is a packet filter (firewall) available in some operating systems like FreeBSD, OpenBSD and macOS. PFLog interface can be used to log packets directly as opposed to [pflogd](https://man.openbsd.org/pflogd). 

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] ~~I have commented my code, particularly in hard-to-understand areas~~
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc`

## How to test this PR locally

An easiest example configuration of pf.conf would be 
```
pass log all
```
Then create `pflog0` interface using `ifconfig` or `ip` commands
After that you can enable pf by running 
```
sudo pfctl -f /etc/pf.conf
```
You should now be able to sniff packets using tcpdump:
```
tcpdump -n -e -ttt -i pflog0
```
## Related issues

Discussed in pull #1362

## Other changes
Migrated `github.com/tsg/gopacket` to `github.com/google/gopacket`

### CODE REVIEW

**Comment by alexmozzhakov** [comment]
I signed the CA

**Comment by alexmozzhakov** [comment]
👍

**Comment by alexmozzhakov** [comment]
@andrewkroh @andresrc anything new? Nobody looked into it for more than a month...

**Comment by alexmozzhakov** [comment]
...

**Comment by marc-gr** [comment]
Hello @alexmozzhakov ! Thanks for the contribution and sorry for the delayed response. Could you rebase the branch, run `mage fmt update` and commit any changed files?

Thanks!

**Comment by alexmozzhakov** [comment]
@marc-gr `mage fmt update` returned `Unknown target specified: update`, but I think it should do the trick

**Comment by marc-gr** [comment]
jenkins run tests

**Comment by marc-gr** [comment]
@alexmozzhakov there are some build issues

```
# github.com/elastic/beats/v7/packetbeat/sniffer
sniffer/afpacket_linux.go:90:18: h.TPacket.SetBPFFilter undefined (type *afpacket.TPacket has no field or method SetBPFFilter)
# github.com/elastic/beats/v7/packetbeat/sniffer
vet: sniffer/afpacket_linux.go:90:19: h.TPacket.SetBPFFilter undefined (type *afpacket.TPacket has no field or method SetBPFFilter)
Error: failed running go vet, please fix the issues reported: running "go vet ./..." failed with exit code 2
```

If you fix them, please be sure to run `make fmt update` (I should have mentioned before that `mage fmt update` should be run inside `./packetbeat` folder, from the root folder please run `make fmt update` instead). 

If you prefer we can do the required steps and commit to this PR, please let us know what you prefer :)

Thanks!

**Comment by alexmozzhakov** [comment]
@marc-gr Hey! I've fixed error and it turned out the ICMP python test started failing. I've investigated the issue and it turns out to be the problem with `Identifier` and `Sequence Number` headers being now included in payloads (which makes them 4 bytes bigger) and human-readable form of ICMP messages has changed from `EchoRequest(0)` and such to just `EchoRequest`. This may (somewhat) break backward-compatibility, so last commit merely a proposition. Another possible solution would be to edit `protos/icmp/icmp.go:317` and `protos/icmp/icmp.go:332` and subtract 4 (const) from length value (there is hard-coded `if len(data) < 4 {` in gopacket). I'm not sure what `(0)` stands for though.

**Comment by marc-gr** [comment]
> @marc-gr Hey! I've fixed error and it turned out the ICMP python test started failing. I've investigated the issue and it turns out to be the problem with `Identifier` and `Sequence Number` headers are included in ICMPv6 payloads as opposed to ICMPv4 (which makes them 4 bytes bigger). Also, human-readable form of ICMP messages has changed from `EchoRequest(0)` and so to just `EchoRequest`.

Sorry it took so long to get back! Went off my radar. Will take a look at it .

**Comment by alexmozzhakov** [comment]
@marc-gr any updates on this one?

File: CHANGELOG.next.asciidoc [hunk 1]
```diff
@@ -851,6 +851,7 @@ same journal. {pull}18467[18467]
   `host` metadata fields when processing network data from network tap or mirror
   port. {pull}19209[19209]
 - Add ECS fields for x509 certs, event categorization, and related IP info. {pull}19167[19167]
+- Add support for pflog packets {pull}20229[20229]
```
Comment (marc-gr) [inline CHANGELOG.next.asciidoc:854]: Would be nice to reorder this one.

**Review by marc-gr — COMMENTED** [review]
Sorry for the late response!

It is looking good, just a couple of notes aside from my comment on the changelog.

- Could you add some sample pcap to have some tests with pflog packets?
- I do not know for sure if changing to google/gopacket can have some unexpected results, maybe @tsg can have a better idea to see if we can go without the fork.
- I think you might need to update the branch again.

Also, let me know if you want me to do the required changes if you do not have the time or if you need help with them. Thanks!

**Comment by alexmozzhakov** [comment]
@marc-gr 
Cheers!

As for library choice I think the main reason was that google's library was premature and some features may have been missing, but @tsg definitely knows best. 

In addition, I had a quick look at tests on the weekend. As far as I remember, to verify my PR works about half a year ago I've run cli version of packetbeat (golang), but for some reason python test won't produce any output for pflog pcap file (works for others tests). Am I missing something? I ran 'make update' on root folder and maybe 'make python-env'. 

Also, if you can't find any pcap for pflog to poke around I can send it next weekend.

Thank you in advance!

**Comment by tsg** [comment]
@alexmozzhakov Thank you for the PR and sorry for the late reply.

I went down the memory lane a bit and it seems that on top of the old gopacket library I've added:

* SetBPFfilter support for af_packet
* A couple of others improvements for af_packet
* Support for saving to pcap files (pcap_dump_open)

I think all of these are now handled in the upstream version as well and you have made the required API changes in the PR. So I think that as long as the tests are passing, we are 👍.

Awesome work!

**Comment by marc-gr** [comment]
Thanks @tsg! 

@alexmozzhakov  To set up the golden test files you need to:

1- Add your pcap file in `packetbeat/tests/system/pcaps`
2- Add a new entry in `packetbeat/tests/system/config/golden-tests.yml` you can use any other as an example. Here you can set any specific config that is required for your test or just pass the reference to the pcap file without any other special configuration.
3- Run `GENERATE=1 mage pythonUnitTest` in the `packetbeat` folder.

If you have any issues, feel free to add the pcap file here and I will help with the changes. Thanks for the work!

**Comment by andrewkroh** [comment]
There may be some changes/enhancements to go-packet that have not been accepted upstream. @marc-gr and @adriansr can you please look over those changes and see what it would take to move to the latest.

**Comment by marc-gr** [comment]
👍  will ensure we are not missing anything before moving forward

**Comment by jlind23** [comment]
@andrewkroh what should we do with this PR? Seems to be inactive for a while now.

**Comment by andrewkroh** [comment]
We have already done the conversion to use github.com/google/gopacket. That should make this PR simpler given that work is already complete. It's seems like the main difference here is switch from the `pcap` to `pcapgo` package. We'd need to evaluate if that has any side-effects.

If anyone wants to pursue this please open up a new PR based on the latest `main` that is using `github.com/google/gopacket`.

**Pull request closed without merging**